### PR TITLE
feat: Add iterator to BlobTransactionSidecar

### PR DESCRIPTION
**Pull request by DoTheBestToGetTheBest**
closes https://github.com/alloy-rs/alloy/issues/1255

### CODE REVIEW

File: crates/eips/src/eip4844/sidecar.rs [hunk 1]
```diff
@@ -32,6 +42,152 @@ pub struct BlobTransactionSidecar {
     pub proofs: Vec<Bytes48>,
 }
 
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
+#[repr(C)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[allow(missing_docs)]
+pub struct BlobTransactionSidecarItem {
```
Comment (mattsse) [inline crates/eips/src/eip4844/sidecar.rs:49]: this should use the same serde attributes

File: crates/eips/src/eip4844/sidecar.rs [hunk 2]
```diff
@@ -32,6 +42,152 @@ pub struct BlobTransactionSidecar {
...
+#[derive(Debug)]
+pub enum BlobValidationError {
+    IndexMismatch { expected: usize, found: usize, details: String },
+    HashMismatch { expected: [u8; 32], found: [u8; 32], details: String },
+    ProofVerificationFailed(String),
+    ProofVerificationError(String),
+    InvalidBlobData(String),
+    InvalidCommitmentData(String),
+    InvalidProofData(String),
+}
```
Comment (mattsse) [inline crates/eips/src/eip4844/sidecar.rs:134]: I don't think we need those, the existing error type should be sufficient

File: crates/eips/src/eip4844/sidecar.rs [hunk 3]
```diff
@@ -32,6 +42,152 @@ pub struct BlobTransactionSidecar {
...
+
+/// A Blob hash
+#[derive(Default, Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct IndexedBlobHash {
+    /// The index of the blob
+    pub index: usize,
+    /// The hash of the blob
+    pub hash: B256,
+}
```
Comment (mattsse) [inline crates/eips/src/eip4844/sidecar.rs:177]: this we don't need, there's a HashNum type already

File: crates/eips/src/eip4844/sidecar.rs (hunk 1 above)
Comment (mattsse) [inline crates/eips/src/eip4844/sidecar.rs:49]: this type should be moved below the existing sidecar impl

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by DoTheBestToGetTheBest** [comment]
hey matts thanks for the reivew, made all changes you requested!

File: crates/eips/src/eip4844/sidecar.rs [hunk 4]
```diff
@@ -285,9 +315,96 @@ impl From<c_kzg::Error> for BlobTransactionValidationError {
...
+#[cfg(feature = "serde")]
+fn parse_u64_string<'de, T, D>(de: D) -> Result<T, D::Error>
+where
+    D: serde::Deserializer<'de>,
+    T: FromStr,
+    <T as FromStr>::Err: core::fmt::Display,
+{
+    String::deserialize(de)?.parse().map_err(serde::de::Error::custom)
+}
```
Comment (mattsse) [inline crates/eips/src/eip4844/sidecar.rs:403]: this is just alloy-serde::quantity

File: crates/eips/src/eip4844/sidecar.rs [hunk 5]
```diff
@@ -285,9 +315,96 @@ impl From<c_kzg::Error> for BlobTransactionValidationError {
...
 #[cfg(test)]
 mod tests {
     use super::*;
+
```
Comment (mattsse) [inline crates/eips/src/eip4844/sidecar.rs:407]:
```suggestion
```

File: crates/eips/src/eip4844/sidecar.rs [hunk 6]
```diff
@@ -32,6 +49,19 @@ pub struct BlobTransactionSidecar {
...
+        self.blobs.iter().zip(&self.commitments).zip(&self.proofs).enumerate().map(
+            |(index, ((blob, commitment), proof))| BlobTransactionSidecarItem {
+                index,
+                blob: *blob,
+                kzg_commitment: *commitment,
+                kzg_proof: *proof,
+            },
+        )
+    }
+}
```
Comment (mattsse) [inline crates/eips/src/eip4844/sidecar.rs:64]: this should be an IntoIter impl instead

File: crates/eips/src/eip4844/sidecar.rs [hunk 7]
```diff
@@ -1,17 +1,34 @@
...
+use alloy_rlp::{Decodable, Encodable};
+#[cfg(feature = "kzg")]
+use c_kzg::KzgProof;
+#[cfg(feature = "serde")]
+use core::str::FromStr;
+#[cfg(feature = "serde")]
+use serde::Deserialize;
+#[cfg(feature = "kzg")]
+use sha2::{Digest, Sha256};
+#[cfg(feature = "kzg")]
```
Comment (mattsse) [inline crates/eips/src/eip4844/sidecar.rs:29]: none of these should be necessary, this makes it very hard to read.

please use full paths instead, this way no import is required

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by DaniPopes** [comment]
Superseded by https://github.com/alloy-rs/alloy/pull/1334

**Pull request closed without merging**